### PR TITLE
Feature/37817 prepare checks for redis

### PR DESCRIPTION
**Pull request by GuyHarwood**
- [x] prepare-check service to store in redis
- [x] preserve existing default implementation
- [x] add jest for easy watch based testing
- [x] feature toggle to opt between prepared checks in table storage or redis (defaults to table storage)

### CODE REVIEW

File: admin/services/prepare-check.service.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    config: check.config,
+    createdAt: new Date(),
+    pinExpiresAt: moment(check.pupil.pinExpiresAt).toDate(),
+    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAt'], check.pupil),
+    pupilId: check.pupil.id,
+    questions: check.questions,
+    school: check.school,
+    schoolId: check.school.id,
+    tokens: check.tokens,
+    updatedAt: new Date()
```
Comment (jon-shipley) [inline admin/services/prepare-check.service.js]: Would be good to have the 'valid from' information.  We have `pinExpiresAt`, so perhaps call it `pinValidFrom`?

File: admin/services/redis-cache.service.js [hunk 2]
```diff
@@ -87,4 +87,29 @@ redisCacheService.drop = async (caches = []) => {
   return true
 }
 
+/**
+ * @param {[{key:string, value:object}]} batch a dictionary of items to add to redis
+ * @param {number} ttl time to expiry in seconds, optional.
+ * @returns {Promise<void>}
+ */
+redisCacheService.addBatch = async (batch, ttl = 0) => {
```
Comment (jon-shipley) [inline admin/services/redis-cache.service.js]: I think it's more accurate if ttl defaults to undefined or null if not set.  We might actually want to set the ttl to 0.

File: admin/services/prepare-check.service.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      const check = checks[index]
+      const preparedCheck = constructPreparedCheck(check)
+      const cacheKey = buildKey(check.schoolPin, check.pupilPin)
+      batch.push({
+        key: cacheKey,
+        value: preparedCheck
+      })
+      const ttl = secondsBetweenNowAnd4pm()
+      return redisService.addBatch(batch, ttl)
+    }
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/services/prepare-check.service.js]: much cleaner :) thanks for the example code 👍

File: admin/services/prepare-check.service.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    return redisService.setMany(preparedChecks)
+  }
+}
+
+/**
+ *
+ * @param {Date} pinExpiry
+ * @returns {number} number of seconds between now and the pin expiry time
+ */
+function secondsBetweenNowAndPinExpiryTime (pinExpiry) {
```
Comment (jon-shipley) [inline admin/services/prepare-check.service.js]: If pinExpiry is coming from another object it is very likely a Moment.moment object

File: admin/services/prepare-check.service.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    createdAt: new Date(),
+    pinExpiresAt: moment(check.pupil.pinExpiresAt).toDate(),
+    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAt'], check.pupil),
+    pupilId: check.pupil.id,
+    questions: check.questions,
+    school: check.school,
+    schoolId: check.school.id,
+    tokens: check.tokens,
+    updatedAt: new Date(),
+    pinValidFrom: new Date()
```
Comment (jon-shipley) [inline admin/services/prepare-check.service.js]: Pins are only valid from some known time in the morning. 8AM?  This time should be local time for the school's timezone.

Comment (GuyHarwood) [reply admin/services/prepare-check.service.js]: sounds like new functionality is being introduced.
This will need to be raised with the PO

Comment (sinhaamit1980) [reply admin/services/prepare-check.service.js]: This is what Paul was saying when we were discussing about MOD schools . He says even though we allowed to generate pin before 8 AM...but those pin should be valid from 8:00AM. As far as i know as soon as we generate pin, pupil can login from that point.

Comment (jon-shipley) [reply admin/services/prepare-check.service.js]: fair enough

Comment (GuyHarwood) [reply admin/services/prepare-check.service.js]: I’ll remove this in light of what we discussed earlier.  It has its own ticket now

Comment (GuyHarwood) [reply admin/services/prepare-check.service.js]: `pinValidFrom` removed, and `updatedAt` and `createdAt` set to moments.

**Pull request merged**